### DISCUSSION TITLE
Simulating long degradation experiments by using Okane 2022 dataset

### DISCUSSION THREAD

**Discussion by benshang1997** [pybamm-team/PyBaMM#4344] [Q&A]
Hello everyone，

I am doing some research about fast charging optimization. I need save pybamm models in different degradation stages from a long  degradation experiments.

I create a DFN Model coupled degradation mechanisms using Okane2022, it is just as same as the pybamm document.

However I find that the degradation speed is very slow. According to the result, the original capacity of battery is 5.109Ah, and the capacity after 200 cycles  is 5.04, with just 0.06 reduced, but it cost nearly 3 hours to simulate. And my final aim is to save the model under 80% soc ( 3Ah). and obviously, the time cost would be too expensive

<img width="1076" alt="1723726335258" src="https://github.com/user-attachments/assets/5a3244bb-24e0-457a-af33-534e337308f4">

so  I want to ask: 1. Is the degradation model correct? Is the degradation speed so slow under Okane2022 set?
                            2.how should i do if i want to speed the simulation? 
                            3. Is there any other way to get the  pybamm models in different degradation stages  except run and save a long degradation experiment.

please find my code attached. thank you very much!

import pybamm
import matplotlib as mpl
mpl.use('tkagg')
import multiprocessing

pybamm.set_logging_level("NOTICE")
model = pybamm.lithium_ion.DFN(
    {
        "thermal": "lumped",
        "SEI": "solvent-diffusion limited",
        "SEI porosity change": "false",
        "lithium plating": "partially reversible",
        "lithium plating porosity change": "false",  # alias for "SEI porosity change"
        "particle mechanics": ("swelling and cracking", "swelling only"),
        "SEI on cracks": "true",
        "loss of active material": "stress-driven",
        "calculate discharge energy": "true",
    })
parameter_values = pybamm.ParameterValues("OKane2022")
# Calculate stoichiometries at 100% SOC
parameter_values.update({
    "Initial temperature [K]": 298.15,  
    "Ambient temperature [K]": 298.15, 
    "Total heat transfer coefficient [W.m-2.K-1]": 5,
    "Upper voltage cut-off [V]": 4.5, },
    check_already_exists=False, )

def train(age_stage):
    print("Training started with age_stage: ", age_stage)
    experiment = pybamm.Experiment(
        [
            "Discharge at 1C until 3.0 V",
            "Rest for 1 hour",
            "Charge at 1 C until 4.1 V",
            "Hold at 4.1 V until 50 mA",
            "Rest for 1 hour",
        ] * 100000, termination=f"{age_stage} Ah capacity"
    )
    sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=experiment)
    safe_solver = pybamm.CasadiSolver(extra_options_setup={"max_step_size": "1e-3"}, atol=1e-8, rtol=1e-8)
    sim.solve(solver=safe_solver, initial_soc=0.2)

    # 导出数据
    sim.save_model(f"sim_model_variables_{age_stage} ", variables=True)

if __name__ == "__main__":
    # Parallel computing
    age_stage = [4.75, 4.5,4.25,4.0,3.75,3.5]
    cpu_cnt = multiprocessing.cpu_count()
    print(f"CPU Count: {cpu_cnt}")
    pool = multiprocessing.Pool()
    pool.map(train,age_stage)

    pool.close()
    pool.join()

**Comment by valentinsulzer** [comment]
The easiest way to speed it up is to use SPMe instead of DFN. At 1C there should not be much difference.

**Reply by benshang1997** [reply]
thank you very much for your quick reply. and may i confirm if it is  right that the degradation is very slow under the condition of Okane 2022 param set?

**Reply by valentinsulzer** [reply]
It's possible ... @DrSOKane ?

**Reply by DrSOKane** [reply]
For OKane2022, SEI porosity change is *true* and thermal should be *isothermal*. You do not need to calculate the discharge energy either, but that doesn't affect degradation.

**Reply by benshang1997** [reply]
Hi @DrSOKane ,thanks for your reply. your work is very excellent. I am a starter on pybamm. In my research, I try to optimize charging  strategy for battery in different aging stage。 so I need to get the battery model with different degradation. My method is to set the termination step as the ageing capacity and run the pybamm to save the model by using function "sim.save_model"，Do you think if this idea is feasible？ And I found the capacity has only decreased by 0.1 after thousands cycles， Is this a normal phenomenon because In other datasets, I found this only requires a few cycles under other datasets, thank you very much.

**Reply by benshang1997** [reply]
Hi @valentinsulzer，thanks for your suggestion, the simulation speed was improved after using SPMe, but a new issues occured: I set the termination as 4.9 Ah, means the simulation will be ended when the capacity reduced to 4.75. I use function"Notice" to monitor this progress. the print just like this:

 2024-08-16 21:48:07.738 - [NOTICE] logger.func(15): Capacity is now 5.109 Ah (originally 5.110 Ah, will stop at 4.900 Ah)

the print message is right. And the simulation end up since capacity (4.900 Ah) is below stopping capacity (4.900 Ah).but the NOTICE message shows the capacity is now 5.03Ah. it puzzled me. and I am not sure where is wrong.

<img width="850" alt="1723844625885" src="https://github.com/user-attachments/assets/7a9c1e3e-044d-4b63-a0fc-2b2ff3d7e600">

**Reply by DrSOKane** [reply]
> Hi @DrSOKane ,thanks for your reply. your work is very excellent. I am a starter on pybamm. In my research, I try to optimize charging strategy for battery in different aging stage。 so I need to get the battery model with different degradation. My method is to set the termination step as the ageing capacity and run the pybamm to save the model by using function "sim.save_model"，Do you think if this idea is feasible？ And I found the capacity has only decreased by 0.1 after thousands cycles， Is this a normal phenomenon because In other datasets, I found this only requires a few cycles under other datasets, thank you very much.

Sorry, I misunderstood the first time. I thought you were trying to reproduce my results, but you're actually running a different experiment with a smaller voltage range. You are cycling between 4.1 V and 3.0 V, whereas I cycled between 4.2 V and 2.5 V. The voltage range has a big impact on the stress-driven loss of active material model.

Most importantly, you should be using `"SEI porosity change": "true"` if you are using DFN or SPMe. It's only if you're using SPM that is should be `"false"`. Try changing this - and nothing else - to see what happens.

Best of luck, Simon

**Reply by benshang1997** [reply]
got it ，thank you very much

**Reply by benshang1997** [reply]
> > Hi @DrSOKane ,thanks for your reply. your work is very excellent. I am a starter on pybamm. In my research, I try to optimize charging strategy for battery in different aging stage。 so I need to get the battery model with different degradation. My method is to set the termination step as the ageing capacity and run the pybamm to save the model by using function "sim.save_model"，Do you think if this idea is feasible？ And I found the capacity has only decreased by 0.1 after thousands cycles， Is this a normal phenomenon because In other datasets, I found this only requires a few cycles under other datasets, thank you very much.
> 
> Sorry, I misunderstood the first time. I thought you were trying to reproduce my results, but you're actually running a different experiment with a smaller voltage range. You are cycling between 4.1 V and 3.0 V, whereas I cycled between 4.2 V and 2.5 V. The voltage range has a big impact on the stress-driven loss of active material model.
> 
> Most importantly, you should be using if you are using DFN or SPMe. It's only if you're using SPM that is should be . Try changing this - and nothing else - to see what happens.`"SEI porosity change": "true"``"false"`
> 
> Best of luck, Simon

Sorry Simon，I didn't understand the meaning of this sentence “ you should be using if you are using DFN or SPMe. It's only if you're using SPM that is should be .” what should I use in DFN and SPMe? thanks

**Reply by DrSOKane** [reply]
In the options dictionary, you shoud be using `"SEI porosity change": "true"`. At the moment, you are using `"SEI porosity change": "false"`, which is incorrect.

**Comment by Florana1** [comment]
Hi, l saw your termination condition is a threshold of capacity. l am wondering how to get the capacity in each cycle? l also want to know the capacity change in the ageing cycles. Thanks in advance.
